### PR TITLE
Allow users to filter mentors by availability

### LINKED ISSUE #182: Add 'isAvailable' to mentors' filter
Currently the API returns also the unavailable mentors. In most of the cases it useless because mentees looking for an available mentor.
Having said that, maybe the user would want to search for someone who is not available anymore by some reason, we can let them to do this.

Reference:
https://github.com/Coding-Coach/find-a-mentor-api//blob/c62d174ec69a7b36ed9128757b6c33a1625eca44/src/modules/common/dto/mentorfilters.dto.ts#L5

### PR DESCRIPTION
**Pull request by arku**
Fixes #182

- Adds E2E tests
- Updates the Swagger doc

### CODE REVIEW

File: src/modules/common/dto/mentorfilters.dto.ts [hunk 1]
```diff
@@ -1,12 +1,14 @@
 import { ApiModelPropertyOptional } from '@nestjs/swagger';
 import { IsOptional } from 'class-validator';
+import { Transform } from 'class-transformer';
 import { PaginationDto } from './pagination.dto';
 
 export class MentorFiltersDto extends PaginationDto {
-  // Private fields used internally only
+  @ApiModelPropertyOptional()
+  @IsOptional()
+  @Transform((val: string) => val === 'true')
```
Comment (arku) [inline src/modules/common/dto/mentorfilters.dto.ts:9]: Let me know if we should change this.

File: src/modules/common/dto/pagination.dto.ts [hunk 2]
```diff
@@ -17,6 +17,10 @@ export class PaginationDto {
   readonly hasMore: boolean;
 
   constructor(values) {
+    if (!values) {
+      return;
+    }
```
Comment (arku) [inline src/modules/common/dto/pagination.dto.ts:22]: I had to add this early return since `ValidationPipe` instantiates a new `MentorFiltersDto` object without passing any values.

**Review by brentmclark — APPROVED** [review]
Overall it looks good to me.  I'd love to learn more about that `@IsBoolean` topic when you get a chance.

Also, my comment marked with a 🐼  is not a request for change.  I tend to favor readability in code, which I understand is subjective.  Future maintainers of this code base will likely be people growing into the backend space, so I favor readability even more so on the CodingCoach repositories.

That said, I don't think the concept of code readability should get in the way of deliver value to our community, especially not when we're talking about 3 lines of code.  

tl;dr:  Feel free to disagree and merge anyway 😄

File: src/modules/common/dto/mentorfilters.dto.ts (hunk 1 above)
Comment (arku) [reply src/modules/common/dto/mentorfilters.dto.ts:9]: The client sends the `available` param in the query string — it'll be a string. I had to explicitly add the transform to convert it to a boolean. If we add `@IsBoolean`, a request to `/mentors?available=true` will fail with a 400.

Do let me know if I'm missing something.

File: src/modules/common/dto/pagination.dto.ts (hunk 2 above)
Comment (arku) [reply src/modules/common/dto/pagination.dto.ts:22]: If you comment out the early return and run the tests, you'll notice the following error.

```
TypeError: Cannot read property 'page' of undefined
    at new PaginationDto (find-a-mentor-api/src/modules/common/dto/pagination.dto.ts:22:15)
    at new MentorFiltersDto (find-a-mentor-api/src/modules/common/dto/mentorfilters.dto.ts:6:1)
    at TransformOperationExecutor.Object.<anonymous>.TransformOperationExecutor.transform (find-a-mentor-api/src/TransformOperationExecutor.ts:123:32)
    at ClassTransformer.Object.<anonymous>.ClassTransformer.plainToClass (find-a-mentor-api/src/ClassTransformer.ts:43:25)
    at Object.plainToClass (find-a-mentor-api/src/index.ts:37:29)
    at ValidationPipe.transform (find-a-mentor-api/node_modules/@nestjs/common/pipes/validation.pipe.js:55:41)
    at find-a-mentor-api/node_modules/@nestjs/core/pipes/pipes-consumer.js:15:33
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at find-a-mentor-api/node_modules/@nestjs/core/router/router-execution-context.js:137:31
    at async Promise.all (index 0)
    at pipesFn (find-a-mentor-api/node_modules/@nestjs/core/router/router-execution-context.js:134:13)
    at find-a-mentor-api/node_modules/@nestjs/core/router/router-execution-context.js:35:30
    at find-a-mentor-api/node_modules/@nestjs/core/router/router-execution-context.js:44:28
    at find-a-mentor-api/node_modules/@nestjs/core/router/router-proxy.js:8:17
    ```

File: src/modules/common/mentors.service.ts [hunk 3]
```diff
@@ -35,6 +35,9 @@ export class MentorsService {
   async findAll(filters: MentorFiltersDto, isLoggedIn: boolean): Promise<any> {
     const onlyMentors: any = {
       roles: Role.MENTOR,
+      ...('available' in filters && {
+        available: filters.available,
+      }),
     };
```
[the remark this answers is not in the document]
Comment (arku) [reply src/modules/common/mentors.service.ts:41]: > What are your thoughts on intentionally making this code a bit more verbose to aid in readability?

Sure, let me do it. But the condition has to be `'available' in filters`. If we check for `filters.available`, users will never be able to find unavailable mentors.

From https://github.com/Coding-Coach/find-a-mentor-api/issues/182#issue-797769589

> maybe the user would want to search for someone who is not available anymore by some reason, we can let them to do this.

[the remark this answers is not in the document]
Comment (brentmclark) [reply src/modules/common/mentors.service.ts]: No worries.  Thanks for hearing me out!

File: src/modules/common/dto/mentorfilters.dto.ts (hunk 1 above)
Comment (brentmclark) [reply src/modules/common/dto/mentorfilters.dto.ts:9]: Thanks for explaining!  I didn't realize that the value of `available` was the string `true` instead of a primitive boolean `true`.  We can always add a custom validator if we get sick of this string transform problem, but this will definitely work for now.

Thanks again!

File: src/modules/common/dto/pagination.dto.ts (hunk 2 above)
Comment (brentmclark) [reply src/modules/common/dto/pagination.dto.ts:22]: That's a bummer.  Nice work on figuring that out!

**Review by crysfel — APPROVED** [review]
Looking great! Thanks!
